File: B07 Merge/Leetcode 23 Merge K Sorted Lists/Leetcode 23.java
/*


*/ 


/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        if (lists == null || lists.length == 0) return null; 
        
        PriorityQueue<ListNode> minHeap = new PriorityQueue<>((O1, O2) ->O1.val - O2.val );
        for(int i = 0; i < lists.length; i++){
            if(lists[i] != null){
                minHeap.add(lists[i]);
            }
        }
        
        ListNode dummy = new ListNode(-1); 
        ListNode cur = dummy; 
        while(minHeap.size() != 0){
            ListNode node = minHeap.remove(); 
            cur.next = node; 
            cur = cur.next; 
            if(node.next != null) minHeap.add(node.next); 
        }
        return dummy.next; 
    }
}

File: B07 Merge/Leetcode 23 Merge K Sorted Lists/Leetcode 23.py
###########


#############
def mergeKLists_Python2(self, lists):
	h = []
	head = tail = ListNode(0)
	for i in lists:
		if i:
			heapq.heappush(h, (i.val, i))

	while h:
		node = heapq.heappop(h)[1]
		tail.next = node
		tail = tail.next
		if node.next:
			heapq.heappush(h, (node.next.val, node.next))

	return head.next



    ######Simple Heap Solution
    class Solution:
    def mergeKLists(self, lists: List[ListNode]) -> ListNode:
        head = ListNode(None)
        curr = head
        h = []
        for i in range(len(lists)):
            if lists[i]:
                heapq.heappush(h, (lists[i].val, i))
                lists[i] = lists[i].next
        
        while h:
            val, i = heapq.heappop(h)
            curr.next = ListNode(val)
            curr = curr.next
            if lists[i]:
                heapq.heappush(h, (lists[i].val, i))
                lists[i] = lists[i].next
        
        return head.next